core/src/main/java/org/elasticsearch/action/support/replication/TransportReplicationAction.java

```diff
@@ -242,10 +241,177 @@ public void messageReceived(final ReplicaRequest request, final TransportChannel
         }
     }
 
+    class PrimaryOperationTransportHandler implements TransportRequestHandler<Request> {
+
+        @Override
+        public void messageReceived(Request request, final TransportChannel channel) throws Exception {
+            ActionListener<Response> actionListener = new ActionListener<Response>() {
+                @Override
+                public void onResponse(Response response) {
+                    try {
+                        channel.sendResponse(response);
+                    } catch (Throwable t) {
+                        onFailure(t);
+                    }
+                }
+
+                @Override
+                public void onFailure(Throwable e) {
+                    try {
+                        channel.sendResponse(e);
+                    } catch (Throwable t) {
+                        logger.warn("failed to send response for get", t);
+                    }
+                }
+            };
+            new AsyncPrimaryAction(request, actionListener).run();
+        }
+    }
+
+    class AsyncPrimaryAction extends AbstractRunnable {
+
+        private final ClusterStateObserver observer;
+        private final ActionListener<Response> listener;
+        private final InternalRequest internalRequest;
+        private final Request request;
+        private final AtomicBoolean finished = new AtomicBoolean(false);
+        private volatile Releasable indexShardReference;
+
+        public AsyncPrimaryAction(Request request, ActionListener<Response> listener) {
+            this.observer = new ClusterStateObserver(clusterService, request.timeout(), logger);
+            this.listener = listener;
+            this.request = request;
+            this.internalRequest = new InternalRequest(request);
+            this.internalRequest.concreteIndex(request.internalShardRouting.getIndex());
+        }
+
+        @Override
+        public void onFailure(Throwable t) {
+            finishAsFailed(t);
```

Comment:
I am not sure anymore if we can completely remove the retry mechanism from the PrimaryAction. Here is my reasoning, let me know if it makes sense: If we only retry on the coordinating node than we assume that the coordinating node is not up-to-date and might receive a newer cluster state after the failure. However, it might also be for a very short time that the receiving node that has the primary is not up-to-date. This is the case in the test for the endless request loop. In this case the coordinating node might not receive a new cluster state at all and the indexing request would then only be executed when the observer times out. The coordinating node has currently no way of knowing if the receiving node got a new cluster state or not. If we had a retry in the AsyncPrimaryAction too we could catch that maybe?